File: solutions/_1405_longest_happy_string/solution.py
from heapq import heappush, heappop, heapify


class Solution:
    def longestDiverseString(self, a: int, b: int, c: int) -> str:
        max_heap = []
        for count, char in [(-a, "a"), (-b, "b"), (-c, "c")]:
            if count:
                heappush(max_heap, (count, char))
        res = []
        while max_heap:
            countA, charA = heappop(max_heap)
            # Check if the last two characters are the same as the current character
            if len(res) > 1 and res[-1] == res[-2] == charA:
                if not max_heap:
                    break
                # If the current character is the same as the last two characters,
                # pick the next character
                countB, charB = heappop(max_heap)
                res.append(charB)
                countB += 1
                if countB:
                    heappush(max_heap, (countB, charB))
            else:
                res.append(charA)
                countA += 1
            if countA:
                heappush(max_heap, (countA, charA))

        return "".join(res)


# https://leetcode.com/problems/longest-happy-string/

File: solutions/_1405_longest_happy_string/solution.js
const { PriorityQueue } = require("datastructures-js");

/**
 * @param {number} a
 * @param {number} b
 * @param {number} c
 * @return {string}
 */
var longestDiverseString = function (a, b, c) {
  const pq = new PriorityQueue((A, B) => B[0] - A[0]);
  for (const [count, char] of [
    [a, "a"],
    [b, "b"],
    [c, "c"],
  ]) {
    if (count > 0) {
      pq.enqueue([count, char]);
    }
  }

  const res = [];
  while (!pq.isEmpty()) {
    let [countA, charA] = pq.dequeue();
    if (
      res.length >= 2 &&
      res[res.length - 1] === charA &&
      res[res.length - 2] == charA
    ) {
      if (pq.isEmpty()) break;
      let [countB, charB] = pq.dequeue();
      res.push(charB);
      countB--;
      if (countB) {
        pq.enqueue([countB, charB]);
      }
    } else {
      res.push(charA);
      countA--;
    }
    if (countA) {
      pq.enqueue([countA, charA]);
    }
  }

  return res.join("");
};

module.exports = { longestDiverseString };

/*
https://leetcode.com/problems/longest-happy-string/
*/
